mllib/src/main/scala/org/apache/spark/ml/Transformer.scala

```diff
@@ -79,7 +80,7 @@ abstract class Transformer extends PipelineStage {
  * result as a new column.
  */
 @DeveloperApi
-abstract class UnaryTransformer[IN, OUT, T <: UnaryTransformer[IN, OUT, T]]
+abstract class UnaryTransformer[IN: TypeTag, OUT: TypeTag, T <: UnaryTransformer[IN, OUT, T]]
```

Comment:
`TypeTag` is required for typed UDF when create udf for `createTransformFunc`.

Reply:
This is a breaking change, but I think it's better than silent result changing.

Reply:
We can avoid this breaking change if we know that the type parameter won't be primitive types. cc @srowen @zhengruifeng

Reply:
I don't disagree, but this is trading a possible error for a definite error. In light of the recent conversations about not-breaking things, is this wise? (I don't object though.)

Yes, let's restrict this to primitive types. I think Spark ML even uses some UDFs that accept AnyRef or something to work with tuples or triples, IIRC.

Reply:
This is a developer API, so I'm wondering if third-party implementations would use primitive type and hit the silent result changing.

I think it's better to ask users to re-compile their Spark application than telling them that they may hit result changinng.